### DISCUSSION TITLE
ARX-Issue-377

### DISCUSSION THREAD

**Discussion by idhamari** [iaBIH/arx#18] [ARX Issues]
[[FEATURE/BUG] Enable definition of training and validation set for privacy-preserving machine learning](https://github.com/arx-deidentifier/arx/issues/377):

The privacy-preserving machine learning framework in ARX uses k-fold cross-validation to quantify the performance of privacy-preserving models. This can lead to misleading estimates as training and validation set both influences the optimization process performed during anonymization.

As an alternative, it would be good to enable users to specify training and a validation set in such a way that only the training set influences the anonymization process. This can easily be done in ARX by using the "research subset" feature, which allows selecting a subset of the records in a dataset that is then anonymized. What would be needed to be added is a feature that allows specifying that machine learning performance is determined based on the set of records that is not included in the research subset.

**Comment by idhamari** [comment]
Reproduction steps:
As described in page 47, Task 5 in ARX workshop. 
1. Open 'aufgabe-5.deid' 
2. Navigate to "Analyze utility" then select "Classification models" 
3. Select a classifier e.g. Logistic regression
4. Select "Feature variables" e.g. "sex, age, education, salary-class " and "Target variables" e.g. "martial-status"
5. The accuracy of the anonymized data is less than the original data. (Accuracy - Baseline accuracy) 
![Screenshot from 2022-03-01 13-50-33](https://user-images.githubusercontent.com/28802516/156172350-507a28db-e13f-4d24-9d9e-97a94184eb1c.png)

**The problem:** the evaluation above is not based on separated test data (It uses all datasets rows for training which allows the classifier to memorize the data (overfitting)). 

Another issue is the computation happens online (when someone change the features or the classifier, the computation starts). It would be nice to add a button to start/stop classification. 

**Note:**
 - The k-fold cross validation part can be found in [arx/aggregates/StatisticsClassification.java](https://github.com/iaBIH/arx/blob/31a409f859479468a3545104c06be91d26261e71/src/main/org/deidentifier/arx/aggregates/StatisticsClassification.java#L364)
- This part should be modified to include training-data and test-data instead of all dataset.

**Comment by idhamari** [comment]
**Notes:**

* Papers related: 
  -  [F. Prasser, J. Eicher, R. Bild, H. Spengler and K. A. Kuhn, "A Tool for Optimizing De-identified Health Data for Use in Statistical Classification," 2017 IEEE 30th International Symposium on Computer-Based Medical Systems (CBMS), 2017, pp. 169-174, doi: 10.1109/CBMS.2017.105.](https://ieeexplore.ieee.org/document/8104179). The paper discussion can be found [here](https://github.com/iaBIH/arx/discussions/14).
* Classification files can be found in:
  *  Arx core:  
      - [/arx/aggregates/StatisticsClassification.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/StatisticsClassification.java) 
      - [/arx/aggregates/ClassificationConfigurationRandomForest.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/ClassificationConfigurationRandomForest.java) 
      - [/arx/aggregates/ClassificationConfigurationNaiveBayes.java ](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/ClassificationConfigurationNaiveBayes.java) 
      - [/arx/aggregates/ClassificationConfigurationLogisticRegression.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/ClassificationConfigurationLogisticRegression.java) 
      -  [/arx/aggregates/classification/MultiClassZeroRClassificationResult.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassZeroRClassificationResult.java)
      -  [/arx/aggregates/classification/MultiClassZeroR.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassZeroR.java)
      -  [/arx/aggregates/classification/MultiClassRandomForestClassificationResult.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassRandomForestClassificationResult.java)
      -  [/arx/aggregates/classification/MultiClassRandomForest.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassRandomForest.java)
      -  [/arx/aggregates/classification/MultiClassNaiveBayesClassificationResult.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassNaiveBayesClassificationResult.java)
      -  [/arx/aggregates/classification/MultiClassNaiveBayes.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassNaiveBayes.java)
      -  [/arx/aggregates/classification/MultiClassLogisticRegressionClassificationResult.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassLogisticRegressionClassificationResult.java)
      -  [/arx/aggregates/classification/MultiClassLogisticRegression.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/MultiClassLogisticRegression.java)
      -  [/arx/aggregates/classification/ClassificationResult.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/ClassificationResult.java)
      -  [/arx/aggregates/classification/ClassificationMethod.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/ClassificationMethod.java)
      -  [/arx/aggregates/classification/ClassificationFeatureMetadata.java](https://github.com/iaBIH/arx/tree/master/src/main/org/deidentifier/arx/aggregates/classification/ClassificationFeatureMetadata.java)
      - [arx/algorithm/LIGHTNINGAlgorithm.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/algorithm/LIGHTNINGAlgorithm.java)
      - [/arx/metric/v2/MetricSDClassification.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/metric/v2/MetricSDClassification.java)
   * Arx GUI:  
      - [gui/view/impl/menu/DialogClassificationConfiguration.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/menu/DialogClassificationConfiguration.java)
      - [gui/view/impl/utility/ViewStatisticsClassificationConfiguration.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/utility/ViewStatisticsClassificationConfiguration.java)      
      - [gui/view/impl/utilityViewStatisticsClassification.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/utility/ViewStatisticsClassification.java)
      - [gui/view/impl/utilityViewStatisticsClassificationAttributes.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/utility/ViewStatisticsClassificationAttributes.java)
      - [gui/view/impl/define/ViewAttributeList.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/define/ViewAttributeList.java)
      - [gui/view/impl/common/ViewData.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/common/ViewData.java)
      - [gui/view/impl/common/ViewDataInput.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/common/ViewDataInput.java)
      - [gui/view/impl/common/ViewDataOutput.java](https://github.com/iaBIH/arx/tree/master/src/gui/org/deidentifier/arx/gui/view/impl/common/ViewDataOutput.java)
   * ARX test:  
      - [TestClassification.java](src/test/org/deidentifier/arx/test/TestClassification.java)

**Reply by idhamari** [reply]
@prasser

**The problem:**
We don't want to use the complete input dataset for finding the anonymization transformation rules.

Why? 
* In a real-world application, the current dataset may not represent the whole population of the dataset and consider only samples. 
* Opinion: The solution: Because in the real-world you will anonymize a (training) dataset without knowing the dataset you will eventually try to classify (with a model trained on the anonymized data).

Suggested Solution:

  1. before anonymizing divide the dataset into training e.g. 80% and testing 20%
  2. run the anonymisation on 80% part  
  3. apply the rules from 2 on 20% part
  4. do classification 80%  for training and 20% for testing  (don't use cross-validation)

**Reply by idhamari** [reply]
Current idea:

Modify StatisticsClassification class to have two functions:

               if allInputIsUsed: 
                      runCrossValidation() // all input records are used in the anonymization
              else   
                   get the remaining records and use for validation
                   runValidation() // only percentage x is used in the anonymisation

**Comment by idhamari** [comment]
**Backtrace:**

* Computing the k-fold cross-validation:  find how the preprocessing and the anonymised data classification work
   1. StatisticsClassification.java [StatisticsClassification](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/main/org/deidentifier/arx/aggregates/StatisticsClassification.java#L315)
   2. StatisticsBuilder.java [getClassificationPerformance](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/main/org/deidentifier/arx/aggregates/StatisticsBuilder.java#L114)
   3. StatisticsBuilderInterruptible.java [getClassificationPerformance](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/main/org/deidentifier/arx/aggregates/StatisticsBuilderInterruptible.java#L106)
   4. ViewStatisticsClassification.java [run](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/gui/org/deidentifier/arx/gui/view/impl/utility/ViewStatisticsClassification.java#L1101)



* Getting random samples of the input data
  1. Controller.java [actionSubsetRandom()](https://github.com/arx-deidentifier/arx/blob/588e1d59041605375c4539b0eb93ab5ba1c83a29/src/gui/org/deidentifier/arx/gui/Controller.java#L1974)
      - Note: it seems it is not implemented in the ARX core 
      - What is the difference between these two lines? using the name model alone is confusing: privacy model? classification model? 
      
                         model
                         this.model

  2. ViewSubsetDefinition.java [build()](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/gui/org/deidentifier/arx/gui/view/impl/define/ViewSubsetDefinition.java#L133)      
  3. [LayoutDefinition.java](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/gui/org/deidentifier/arx/gui/view/impl/define/LayoutDefinition.java#L44)
  4. [MianWindow.java](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/gui/org/deidentifier/arx/gui/view/impl/MainWindow.java#L168)

**Reply by prasser** [reply]
Subsets are fully supported in the ARX Core

// Define subset
DataSubset subset = DataSubset.create(data, new HashSet<Integer>(Arrays.asList(1, 2, 5)));  

// Only subset
config.addPrivacyModel(new Inclusion(subset));  

// Privacy model with subset kmap
config.addPrivacyModel(new KMap(3, subset));  

// Privacy model with subset: delta presence
config.addPrivacyModel(new DPresence(1d / 2d, 2d / 3d, subset));  

// Obtain subset

// Overall dataset
DataHandle handle = result.getOutput();

// Subset
DataHandle subset = handle.getView();

// Check whether handle is subset
if (handle instanceof DataHandleSubset) {
    // Subset
}

**Reply by idhamari** [reply]
Thanks @prasser for your comment. I am reading now the related data class files. This [function](https://github.com/arx-deidentifier/arx/blob/588e1d59041605375c4539b0eb93ab5ba1c83a29/src/gui/org/deidentifier/arx/gui/Controller.java#L1974) is implemented in the gui. Maybe it should be implemented in the [DataSubset](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/main/org/deidentifier/arx/DataSubset.java)

**Reply by prasser** [reply]
Might be convenient for users, but not really needed. A random subset can be implemented trivially by providing random indices using:

DataSubset.create(data, new HashSet(Arrays.asList(1, 2, 5)));

**Comment by idhamari** [comment]
* creating a new examples from [Example39.java](https://github.com/arx-deidentifier/arx/blob/master/src/example/org/deidentifier/arx/examples/Example39.java) to test the new solution:
    - add random sampling  
    - create classification model 
     - apply classification on the non-selected records
    - compare results with the previous approach 

**Idea:** create an interface to SMILE:
  -   export data to SMILE format
  -   call SMILE function 
  -   import result to ARX
 
This way:
- no need to create a file for classifications 
- new methods can be integrated easily

**Reply by idhamari** [reply]
* Typo in [DataSubset.java](https://github.com/arx-deidentifier/arx/blob/32f2102f6908053a15a7cdbac4577ea575d3c1f8/src/main/org/deidentifier/arx/DataSubset.java):  

             /**
              * This class represents a the dataset that is to be de-identified 
              * as a subset of the given population table.
              *

* is there a way to convert between subset and data? 
* [Hashcode](https://www.scaler.com/topics/hashcode-in-java/) is used for speedup.

**Reply by prasser** [reply]
Typo -> PR against ARX

**Comment by prasser** [comment]
is there a way to convert between subset and data?

No. Data is only used at the very beginning. There is no easy way to go back from DataHandle to Data (and this is not needed).

**Comment by prasser** [comment]
Idea: create an interface to SMILE: -> No. New methods need to be integrated analogously to the other ones.

**Reply by idhamari** [reply]
> New methods need to be integrated analogously to the other ones.

That is what I am trying to understand

**Reply by prasser** [reply]
How this is done should be relatively easy to understand by looking at MultiClassLogisticRegression (online learning) and MultiClassRandomForest (no online learning).

However, to change what is used as a training and test set it is not needed to add or change a classifier....
